src/nnet3/decodable-batch-looped.h

```diff
@@ -0,0 +1,357 @@
+  
+  // Advance one chunk in bacth
+  // The sequence represented by any request in batch may be different with last 
+  // chunk, so we should set state for requests before inferene, and get 
+  // state for requests after inference.
+  void AdvanceChunk(const std::vector<NnetComputeRequest*> &requests);
+
+  void Compute();
+
+  bool Continue();
+  
+private:
+  const DecodableNnetBatchLoopedInfo &info_;
+
+  std::vector<NnetComputerSnapshot> snapshots_;
+
+  // Continue computation or not.
+  // It don't needs lock for multi-thread synchronization,
+  // because the assignment is atomic.
+  bool is_working_;
+  
+  // Queue contains requests waiting for computation.
+  // It needs a mutex to help achieve multi-thread synchronization.
+  // The first chunk is different with other chunks, so needs a 
+  // queue to handle requests of first chunk only.
+  typedef std::chrono::system_clock::time_point TimePoint;
+  typedef std::pair<NnetComputeRequest*, TimePoint> QueueElement;
+  std::queue<QueueElement> queue_; 
+  std::queue<QueueElement> chunk1_queue_;
+  std::mutex mtx_;
+  std::condition_variable condition_variable_;
+
+  // The thread for computation
+  std::thread work_thread_;
+
+  // The vector indicates the matrices stored state are batch first or not.
+  std::vector<bool> batch_first_;
+};
+
+
+class DecodableNnetBatchLoopedOnline : public DecodableInterface, 
+  public NotifiableNnetBatchLooped {
+public:
+  // Constructor.  'input_feature' is for the feature that will be given
+  // as 'input' to the neural network; 'ivector_feature' is for the iVectors
+  // OnlineFeatureInterface *ivector_features);
+  DecodableNnetBatchLoopedOnline(NnetBatchLoopedComputer *computer,
+                                 const TransitionModel &trans_model,
+                                 OnlineFeatureInterface *input_features,
+                                 OnlineFeatureInterface *ivector_features);
+
+  virtual BaseFloat LogLikelihood(int32 subsampled_frame, int32 index);
+
+  void GetOutputForFrame(int32 subsampled_frame, VectorBase<BaseFloat> *output);
+  
+  // note: the frame argument is on the output of the network, i.e. after any
+  // subsampling, so we call it 'subsampled_frame'.
+  virtual bool IsLastFrame(int32 subsampled_frame) const;
+
+  virtual int32 NumFramesReady() const;
+
+  virtual int32 NumIndices() const { return trans_model_.NumTransitionIds(); }
+
+  // this is not part of the standard Decodable interface but I think is needed for
+  // something.
+  int32 FrameSubsamplingFactor() const {
+    return info_.opts.frame_subsampling_factor;
+  }
+
+  virtual void Receive(const CuMatrixBase<BaseFloat> &output) {
+    if (current_log_post_.NumRows() != output.NumRows() ||
+        current_log_post_.NumCols() != output.NumCols())
+      current_log_post_.Resize(output.NumRows(), output.NumCols());
+    current_log_post_.CopyFromMat(output);
+    semaphone_.Signal();
+    return ;
```

Comment:
```suggestion
```

Reply:
I think allways with return in methods/funtions, no matter void or not.

Reply:
Yeah, but `return` is **redundant**. You can keep **consistent** with other functions in kaldi, even your own functions.

The fewer you type, the fewer bugs you will have.

Not every function that you are adding has a `return` at the end.